### PR TITLE
feat: make JSON serialization of referenced pydantic models possible

### PR DESCRIPTION
**Pull request by PrettyWood**
## Change Summary
Allow to customize serialization of fields with `BaseModel` type

```py
from typing import List, Optional

from pydantic import BaseModel


class Address(BaseModel):
    city: str
    country: str


class User(BaseModel):
    name: str
    address: Address
    friends: Optional[List['User']] = None

    class Config:
        json_encoders = {
            Address: lambda a: f'{a.city} ({a.country})',
            'User': lambda u: f'{u.name} in {u.address.city} '
                              f'({u.address.country[:2].upper()})',
        }


User.update_forward_refs()

wolfgang = User(
    name='Wolfgang',
    address=Address(city='Berlin', country='Deutschland'),
    friends=[
        User(name='Pierre', address=Address(city='Paris', country='France')),
        User(name='John', address=Address(city='London', country='UK')),
    ],
)
print(wolfgang.json())
# {"name": "Wolfgang", "address": "Berlin (Deutschland)", "friends": ["Pierre in Paris (FR)", "John in London (UK)"]}
```

## Related issue number
https://github.com/samuelcolvin/pydantic/discussions/2622

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/json.py [hunk 1]
```diff
@@ -98,10 +98,13 @@ def pydantic_encoder(obj: Any) -> Any:
...
     # Check the class type and its superclasses for a matching encoder
     for base in obj.__class__.__mro__[:-1]:
-        try:
+        if base in type_encoders:
             encoder = type_encoders[base]
-        except KeyError:
+        elif base.__name__ in type_encoders:
+            encoder = type_encoders[base.__name__]
+        else:
             continue
```
Comment (samuelcolvin) [inline pydantic/json.py:106]: To make this as performant as possible, I think we should do something like

```py
try:
    encoder = type_encoders[base]
except KeyError:
    try:
        encoder = type_encoders[base.__name__]
    except KeyError:
        continue
```

This avoids duplicated lookups.

File: pydantic/main.py [hunk 2]
```diff
@@ -538,13 +538,15 @@ def json(
...
         encoder = cast(Callable[[Any], Any], encoder or self.__json_encoder__)
-        data = self.dict(
-            include=include,
-            exclude=exclude,
-            by_alias=by_alias,
-            exclude_unset=exclude_unset,
-            exclude_defaults=exclude_defaults,
-            exclude_none=exclude_none,
+        data = dict(
+            self._iter(
```
Comment (samuelcolvin) [inline pydantic/main.py:542]: what does this change do?

How does this change effect performance?

Comment (samuelcolvin) [reply pydantic/main.py:542]: might be worth adding a comment.

**Review by samuelcolvin — COMMENTED** [review]
Seems like a good start.

**Comment by samuelcolvin** [comment]
please update.

**Comment by PrettyWood** [comment]
Thanks for the remarks. Please review the updated version

**Comment by mcauto** [comment]
When is it merging?

**Comment by samuelcolvin** [comment]
thank you.

**Pull request merged**